### ISSUE TITLE
agent activation issue: Unable to run tpm2_policysecret

### ISSUE THREAD

**Issue by stefanberger** [keylime/keylime#584]
# Environment
 * OS / version: Fedora 32
 * Processor architecture: PowerPC64
 * TPM Manufacturer: Nuvoton
 * Keylime version: git master (1766957a748)

# Description 

Start of keylime_agent at some point results in the following issue. Typically a reboot resolves the issue but I don't know whether this requires a reboot. It could be an issue with the TPM 2 itself or an issue related to some cleanup of the TPM 2 not being done.

# keylime_agent              
Using config file /etc/keylime.conf
2021-03-14 18:37:08.880 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.1.3
2021-03-14 18:37:08.995 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.1.3
2021-03-14 18:37:09.514 - keylime.tpm - INFO - Taking ownership with config provided TPM owner password
2021-03-14 18:37:11.906 - keylime.tpm - INFO - TPM Owner password confirmed: keylime
2021-03-14 18:37:11.906 - keylime.tpm - INFO - Flushing old ek handle: 0x81000001
2021-03-14 18:37:18.230 - keylime.tpm - INFO - Flushing old ak handle: /var/lib/keylime/secure/tmp_1g2t4nm
2021-03-14 18:37:23.506 - keylime.tpm - WARNING - TPM randomness not added to system entropy pool: [Errno 22] Invalid argument
2021-03-14 18:37:23.506 - keylime.cloudagent - INFO - Agent UUID: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-14 18:37:23.710 - keylime.registrar_client - INFO - Agent registration requested for D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-14 18:37:23.899 - keylime.tpm - ERROR - Error decrypting AIK: Command: ['tpm2_policysecret', '-S', '/var/lib/keylime/secure/tmp8_tw5j1u', '-c', '0x4000000B', 'keylime'] returned 1, expected 0, output [], stderr [b'ERROR: Could not read session type\n', b'ERROR: Unable to run tpm2_policysecret\n']
2021-03-14 18:37:23.899 - keylime.tpm - ERROR - Command: ['tpm2_policysecret', '-S', '/var/lib/keylime/secure/tmp8_tw5j1u', '-c', '0x4000000B', 'keylime'] returned 1, expected 0, output [], stderr [b'ERROR: Could not read session type\n', b'ERROR: Unable to run tpm2_policysecret\n']
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/keylime/tpm/tpm_main.py", line 803, in activate_identity
    self.__run(["tpm2_policysecret", "-S", sesspath, "-c", "0x4000000B", owner_pw])
  File "/usr/local/lib/python3.8/site-packages/keylime/tpm/tpm_main.py", line 377, in __run
    raise Exception("Command: %s returned %d, expected %d, output %s, stderr %s" % (cmd, code, expectedcode, retout, reterr))
Exception: Command: ['tpm2_policysecret', '-S', '/var/lib/keylime/secure/tmp8_tw5j1u', '-c', '0x4000000B', 'keylime'] returned 1, expected 0, output [], stderr [b'ERROR: Could not read session type\n', b'ERROR: Unable to run tpm2_policysecret\n']
Traceback (most recent call last):
  File "/usr/local/bin/keylime_agent", line 11, in <module>
    load_entry_point('keylime==0.0.0', 'console_scripts', 'keylime_agent')()
  File "/usr/local/lib/python3.8/site-packages/keylime/cmd/agent.py", line 15, in main
    keylime_agent.main()
  File "/usr/local/lib/python3.8/site-packages/keylime/keylime_agent.py", line 552, in main
    raise Exception("Activation failed")
Exception: Activation failed

## Expected behavior vs. actual behavior 

keylime_agent start should always work.

## Steps to reproduce problem
Restart keylime_agent several times and at some point it will fail.

**Comment by stefanberger** [comment]
I tried this again by repeatedly starting the keylime_agent and terminating it once it had fully initialized. Repeating this several times led to a different failure first and then to this failure:

```
root@ithaca-b ~]# keylime_agent 
Using config file /etc/keylime.conf
2021-03-16 08:05:01.373 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.1.3
2021-03-16 08:05:01.484 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.1.3
2021-03-16 08:05:01.883 - keylime.tpm - INFO - Taking ownership with config provided TPM owner password
2021-03-16 08:05:04.308 - keylime.tpm - INFO - TPM Owner password confirmed: keylime
2021-03-16 08:05:04.308 - keylime.tpm - INFO - Flushing old ek handle: 0x81000000
2021-03-16 08:05:10.304 - keylime.tpm - INFO - Flushing old ak handle: /var/lib/keylime/secure/tmpj6ladzk0
2021-03-16 08:05:15.492 - keylime.tpm - WARNING - TPM randomness not added to system entropy pool: [Errno 22] Invalid argument
2021-03-16 08:05:15.492 - keylime.cloudagent - INFO - Agent UUID: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-16 08:05:15.700 - keylime.registrar_client - INFO - Agent registration requested for D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-16 08:05:20.030 - keylime.tpm - INFO - AIK activated.
2021-03-16 08:05:20.044 - keylime.registrar_client - ERROR - Error: unexpected http response code from Registrar Server: 400
2021-03-16 08:05:20.044 - keylime.registrar_client - ERROR - Response code 400: Error: Auth tag ebdb16d95c24018f04a468be6ab56f060c53a0548d34d8387dfb8e05b4f6964cc3ce2d42b1757e4a63aa72225fb10a6f does not match expected value f66fce2b0b611d74c5744f86d7a2b8b7dfdf58b945672e4166e7984922427f81a2090d5dec1fed6dbab5f2a2064cc7dd
Traceback (most recent call last):
  File "/usr/local/bin/keylime_agent", line 11, in <module>
    load_entry_point('keylime==0.0.0', 'console_scripts', 'keylime_agent')()
  File "/usr/local/lib/python3.8/site-packages/keylime/cmd/agent.py", line 15, in main
    keylime_agent.main()
  File "/usr/local/lib/python3.8/site-packages/keylime/keylime_agent.py", line 560, in main
    raise Exception("Registration failed on activate")
Exception: Registration failed on activate
[root@ithaca-b ~]# keylime_agent 
Using config file /etc/keylime.conf
2021-03-16 08:05:29.277 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.1.3
2021-03-16 08:05:29.365 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.1.3
2021-03-16 08:05:29.858 - keylime.tpm - INFO - Taking ownership with config provided TPM owner password
2021-03-16 08:05:32.231 - keylime.tpm - INFO - TPM Owner password confirmed: keylime
2021-03-16 08:05:32.231 - keylime.tpm - INFO - Flushing old ek handle: 0x81000000
2021-03-16 08:05:38.899 - keylime.tpm - INFO - Flushing old ak handle: /var/lib/keylime/secure/tmpp2ek72wm
2021-03-16 08:05:44.516 - keylime.tpm - WARNING - TPM randomness not added to system entropy pool: [Errno 22] Invalid argument
2021-03-16 08:05:44.516 - keylime.cloudagent - INFO - Agent UUID: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-16 08:05:44.662 - keylime.registrar_client - INFO - Agent registration requested for D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-16 08:05:44.773 - keylime.tpm - ERROR - Error decrypting AIK: Command: ['tpm2_policysecret', '-S', '/var/lib/keylime/secure/tmp4gfjfmft', '-c', '0x4000000B', 'keylime'] returned 1, expected 0, output [], stderr [b'ERROR: Could not read session type\n', b'ERROR: Unable to run tpm2_policysecret\n']
2021-03-16 08:05:44.773 - keylime.tpm - ERROR - Command: ['tpm2_policysecret', '-S', '/var/lib/keylime/secure/tmp4gfjfmft', '-c', '0x4000000B', 'keylime'] returned 1, expected 0, output [], stderr [b'ERROR: Could not read session type\n', b'ERROR: Unable to run tpm2_policysecret\n']
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/keylime/tpm/tpm_main.py", line 803, in activate_identity
    self.__run(["tpm2_policysecret", "-S", sesspath, "-c", "0x4000000B", owner_pw])
  File "/usr/local/lib/python3.8/site-packages/keylime/tpm/tpm_main.py", line 377, in __run
    raise Exception("Command: %s returned %d, expected %d, output %s, stderr %s" % (cmd, code, expectedcode, retout, reterr))
Exception: Command: ['tpm2_policysecret', '-S', '/var/lib/keylime/secure/tmp4gfjfmft', '-c', '0x4000000B', 'keylime'] returned 1, expected 0, output [], stderr [b'ERROR: Could not read session type\n', b'ERROR: Unable to run tpm2_policysecret\n']
Traceback (most recent call last):
  File "/usr/local/bin/keylime_agent", line 11, in <module>
    load_entry_point('keylime==0.0.0', 'console_scripts', 'keylime_agent')()
  File "/usr/local/lib/python3.8/site-packages/keylime/cmd/agent.py", line 15, in main
    keylime_agent.main()
  File "/usr/local/lib/python3.8/site-packages/keylime/keylime_agent.py", line 552, in main
    raise Exception("Activation failed")
Exception: Activation failed
```

A restart of tpm2-abrmd did not resolve the issue.
The handles at the time of failure looked like this:

```
[root@ithaca-b ~]# tpm2_getcap handles-transient
[root@ithaca-b ~]# tpm2_getcap handles-persistent
- 0x81000000
- 0x81000001
- 0x81000002
[root@ithaca-b ~]# tpm2_getcap handles-permanent
- 0x40000001
- 0x40000007
- 0x40000009
- 0x4000000A
- 0x4000000B
- 0x4000000C
- 0x4000000D
[root@ithaca-b ~]# tpm2_getcap handles-loaded-session
[root@ithaca-b ~]# tpm2_getcap handles-saved-session
- 0x2000000
- 0x2000001
```

**Comment by maugustosilva** [comment]
@stefanberger many thanks for opening this issue. I was able to reproduce it (a long time ago) but simply forgot to open the issue. Yes, the only way to solve it in my case was also a reboot. I have tried all the cleanup steps which you described here, with the same unsuccessful results.

**Comment by stefanberger** [comment]
Do you happen to remember the manufacturer of the TPM that was in that machine? It sounds like this is a not-so-often occurring event that may be isolated to only some devices rather than a general keylime issue.

**Comment by maugustosilva** [comment]
I remember being able to reproduce this even on swtpm

**Comment by stefanberger** [comment]
I have tried to re-create the issue inside a VM with an attached vTPM (swtpm) but cannot re-create it.

**Comment by MartinMa28** [comment]
Run into this issue inside a VM with swtpm. It happens after a few successful quotes.

**Comment by MartinMa28** [comment]
After reboot, the error is gone for this time.

**Comment by stefanberger** [comment]
I tried it now with a quote. I had to do one successful quote and a couple of restarts of the keylime_agent until the problem hit. I could resolve the issue by running `rm -f /var/lib/keylime/secure/tmp*` on the keylime_agent side.

The 2nd time I did not do a single quote. I just started the keylime_agent and hit the problem. It may have to do with the number of file entries `/var/lib/keylime/secure/tmp*`. Because once I had removed them again, it again worked. Here's the sequence:

```
root@localhost ~]# ls -l /var/lib/keylime/secure/
total 832
-rw-------. 1 root root 1679 Mar 19 15:07 tci_rsa_key
-rw-------. 1 root root 1256 Mar 19 15:26 tmp0mxqjk05
-rw-------. 1 root root 1256 Mar 19 15:22 tmp0zg89x6n
-rw-------. 1 root root 1256 Mar 19 15:19 tmp11kmkevr
-rw-------. 1 root root 1256 Mar 19 15:29 tmp1dh1pele
-rw-------. 1 root root 1256 Mar 19 15:24 tmp3won_z8j
-rw-------. 1 root root 1256 Mar 19 15:25 tmp7ofdbwzu
-rw-------. 1 root root 1256 Mar 19 15:23 tmpg096s44h
-rw-------. 1 root root 1256 Mar 19 15:27 tmphzer81q_
-rw-------. 1 root root 1256 Mar 19 15:23 tmpkv7uy62j
-rw-------. 1 root root 1256 Mar 19 15:21 tmpn42yckyu
-rw-------. 1 root root 1256 Mar 19 15:23 tmpr_c9g3rj
-rw-------. 1 root root 1256 Mar 19 15:22 tmprd_6hn6z
[root@localhost ~]# keylime_agent 
Using config file /etc/keylime.conf
2021-03-19 15:29:21.583 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.3.0
2021-03-19 15:29:21.735 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.3.0
2021-03-19 15:29:22.143 - keylime.tpm - INFO - Taking ownership with config provided TPM owner password
2021-03-19 15:29:24.411 - keylime.tpm - INFO - TPM Owner password confirmed: keylime
2021-03-19 15:29:24.411 - keylime.tpm - INFO - Flushing old ek handle: 0x81000000
2021-03-19 15:29:30.376 - keylime.tpm - INFO - Flushing old ak handle: /var/lib/keylime/secure/tmp1dh1pele
2021-03-19 15:29:51.118 - keylime.tpm - WARNING - TPM randomness not added to system entropy pool: [Errno 22] Invalid argument
2021-03-19 15:29:51.118 - keylime.cloudagent - INFO - Agent UUID: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-19 15:29:51.316 - keylime.registrar_client - INFO - Agent registration requested for D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-19 15:29:55.613 - keylime.tpm - INFO - AIK activated.
2021-03-19 15:29:55.629 - keylime.registrar_client - INFO - Registration activated for agent D432FBB3-D2F1-4A97-9EF7-75BD81C00000.
2021-03-19 15:29:56.821 - keylime.cloudagent - INFO - Starting Cloud Agent on 0.0.0.0:9002 use <Ctrl-C> to stop
2021-03-19 15:29:56.822 - keylime.revocation_notifier - INFO - Waiting for revocation messages on 0mq 127.0.0.1:8992
^C2021-03-19 15:29:59.101 - keylime.cloudagent - INFO - TERM Signal received, shutting down...
^[[A
[root@localhost ~]# keylime_agent 
Using config file /etc/keylime.conf
2021-03-19 15:30:00.605 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.3.0
2021-03-19 15:30:00.696 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.3.0
2021-03-19 15:30:01.100 - keylime.tpm - INFO - Taking ownership with config provided TPM owner password
2021-03-19 15:30:03.335 - keylime.tpm - INFO - TPM Owner password confirmed: keylime
2021-03-19 15:30:03.335 - keylime.tpm - INFO - Flushing old ek handle: 0x81000000
2021-03-19 15:30:09.655 - keylime.tpm - INFO - Flushing old ak handle: /var/lib/keylime/secure/tmp6j56jk7a
2021-03-19 15:31:09.177 - keylime.tpm - WARNING - TPM randomness not added to system entropy pool: [Errno 22] Invalid argument
2021-03-19 15:31:09.177 - keylime.cloudagent - INFO - Agent UUID: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-19 15:31:09.443 - keylime.registrar_client - INFO - Agent registration requested for D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-19 15:31:13.455 - keylime.tpm - INFO - AIK activated.
2021-03-19 15:31:13.469 - keylime.registrar_client - ERROR - Error: unexpected http response code from Registrar Server: 400
2021-03-19 15:31:13.469 - keylime.registrar_client - ERROR - Response code 400: Error: Auth tag ebdb16d95c24018f04a468be6ab56f060c53a0548d34d8387dfb8e05b4f6964cc3ce2d42b1757e4a63aa72225fb10a6f does not match expected value 7db2be701ab642d1d3450783f1c19e298da0ac8e5a156b833998ebd33c9c6374828e3ffbe8ec7bd6830cd16ebf1a7e33
Traceback (most recent call last):
  File "/usr/local/bin/keylime_agent", line 33, in <module>
    sys.exit(load_entry_point('keylime==0.0.0', 'console_scripts', 'keylime_agent')())
  File "/usr/local/lib/python3.9/site-packages/keylime/cmd/agent.py", line 15, in main
    keylime_agent.main()
  File "/usr/local/lib/python3.9/site-packages/keylime/keylime_agent.py", line 543, in main
    raise Exception("Registration failed on activate")
Exception: Registration failed on activate
[root@localhost ~]# ls -l /var/lib/keylime/secure/
total 960
-rw-------. 1 root root 1679 Mar 19 15:07 tci_rsa_key
-rw-------. 1 root root 1256 Mar 19 15:26 tmp0mxqjk05
-rw-------. 1 root root 1256 Mar 19 15:22 tmp0zg89x6n
-rw-------. 1 root root 1256 Mar 19 15:19 tmp11kmkevr
-rw-------. 1 root root 1256 Mar 19 15:29 tmp1dh1pele
-rw-------. 1 root root 1256 Mar 19 15:24 tmp3won_z8j
-rw-------. 1 root root 1256 Mar 19 15:29 tmp6j56jk7a
-rw-------. 1 root root 1256 Mar 19 15:25 tmp7ofdbwzu
-rw-------. 1 root root 1256 Mar 19 15:23 tmpg096s44h
-rw-------. 1 root root 1256 Mar 19 15:27 tmphzer81q_
-rw-------. 1 root root 1256 Mar 19 15:23 tmpkv7uy62j
-rw-------. 1 root root 1256 Mar 19 15:21 tmpn42yckyu
-rw-------. 1 root root 1256 Mar 19 15:23 tmpr_c9g3rj
-rw-------. 1 root root 1256 Mar 19 15:22 tmprd_6hn6z
-rw-------. 1 root root 1256 Mar 19 15:31 tmpucbh3pis
[root@localhost ~]# rm -rf /var/lib/keylime/secure/tmp*
[root@localhost ~]# keylime_agent 
Using config file /etc/keylime.conf
2021-03-19 15:31:30.425 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.3.0
2021-03-19 15:31:30.544 - keylime.tpm - INFO - TPM2-TOOLS Version: 4.3.0
2021-03-19 15:31:30.800 - keylime.tpm - INFO - Taking ownership with config provided TPM owner password
2021-03-19 15:31:32.502 - keylime.tpm - INFO - TPM Owner password confirmed: keylime
2021-03-19 15:31:32.502 - keylime.tpm - INFO - Flushing old ek handle: 0x81000000
2021-03-19 15:31:38.685 - keylime.tpm - INFO - Flushing old ak handle: /var/lib/keylime/secure/tmpucbh3pis
2021-03-19 15:31:53.604 - keylime.tpm - WARNING - TPM randomness not added to system entropy pool: [Errno 22] Invalid argument
2021-03-19 15:31:53.604 - keylime.cloudagent - INFO - Agent UUID: D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-19 15:31:53.767 - keylime.registrar_client - INFO - Agent registration requested for D432FBB3-D2F1-4A97-9EF7-75BD81C00000
2021-03-19 15:31:57.388 - keylime.tpm - INFO - AIK activated.
2021-03-19 15:31:57.404 - keylime.registrar_client - INFO - Registration activated for agent D432FBB3-D2F1-4A97-9EF7-75BD81C00000.
2021-03-19 15:31:58.508 - keylime.cloudagent - INFO - Starting Cloud Agent on 0.0.0.0:9002 use <Ctrl-C> to stop
2021-03-19 15:31:58.509 - keylime.revocation_notifier - INFO - Waiting for revocation messages on 0mq 127.0.0.1:8992
```

**Comment by stefanberger** [comment]
Closing this issue now since per my testing it has been resolved with PR #624 .
